### ISSUE TITLE
how to initialize multi-process consumers?

### ISSUE THREAD

**Issue by FelixSchwarz** [coleifer/huey#351]
I configured the huey consumer to use workers (worker_type = WORKER_PROCESS, workers = N). I can easily do simple computations. Now my question is how I can pass the initial configuration for my workers.

As a trivial example: My workers should connect to a database and they need DB credentials to do so. How can I pass that information to the workers? (The data is not available via python directly, so no "settings.py" or so but often stored in a config file so the worker must parse the file before executing the first task.)

Basically I'm looking for a hook to execute some code before a worker takes any work. How can I do that with huey.

PS: I saw `@huey.pre_execute` and `Huey.register_pre_execute()` but I don't know how I can pass data to these functions/methods.

**Comment by coleifer** [comment]
Why do your workers need to connect to a database?

**Comment by FelixSchwarz** [comment]
For example because they need some data from them :-) But the database was just one example.
For example a task may be to extract some text from a document and to store the text in an elasticsearch server - need elasticsearch hostname, index name etc.  Or the task should generate a report which takes up to a few hours or so (and the result is - stored in the database). 

Also some (older) frameworks rely on global state which needs to be initialized. Basically this is about having control over workers.

**Comment by coleifer** [comment]
> For example because they need some data from them :-)

I don't want to be presumptuous, but I don't get the impression that you understand very well how huey works. The worker process is spawned by the main consumer process and is managed by the consumer. Workers are not stateful. They simply execute tasks.

If you need to connect to elasticsearch to push some data up to it, you would theoretically pass any connection information to the task.

If the task is generating a report and writing to the database, then inside your task function you would open a connection to the database and perform your writes, and close it afterwards.

Like I said, I don't think you understand how huey works.

**Comment by FelixSchwarz** [comment]
> I don't want to be presumptuous, but I don't get the impression that you understand very well how huey works.

That is certainly possible :-) - though I implemented quite some tasks with celery which should be similar in nature.

> If you need to connect to elasticsearch to push some data up to it, you would theoretically pass any connection information to the task.
> If the task is generating a report and writing to the database, then inside your task function you would open a connection to the database and perform your writes, and close it afterwards.

That is certainly a way of approaching it but in my case I think it means a quite a bit of overhead (ElasticSearch is easy because it is just a simple HTTP request). My current use case is to push about 10-50k tasks during peak hours and in the past we achieved measurable gains (I'm adapting existing code which was used in different scenarios) by reusing DB connections (and similar resources).

However based on your reply I'll just try to do everything per task and see how bad the performance hit actually is. If its too bad I'll check if I can either add some quirks to huey (as I did with a multiprocessing-based adaptation) or need to go to a different library.

**Comment by coleifer** [comment]
I think this is a good idea -- I've got a first-cut at the implementation pushed: b05f70777bcede61385ee9bd50f728e0b809f7e7

```python
db_conn = None

@huey.on_startup()
def setup_db_conn():
    global db_conn
    db_conn = psycopg2.connect(database='my_db', ...)

@huey.task()
def write_to_db(data):
    cursor = db_conn.cursor()
    # ...
```

**Comment by FelixSchwarz** [comment]
Thank you - that looks very much like the API I had in mind, The only issue for me is now how to pass data to the `.on_startup()` function. Probably the `huey` instance could store that data but I think I have an idea at least for my immediate needs.

**Comment by coleifer** [comment]
How would you pass information to it, given that it's called from within the consumer and not by the user?
